### PR TITLE
OTEL import/export

### PR DESCRIPTION
**Pull request by sfc-gh-pmardziel**
# Description

First a note about how the new files in this PR are named/located. I attempted to not modify existing files unless it was not plausible and put all of the experimental feature code in `trulens.experimental.otel_tracing`. The folder structure there is the same as `trulens` meaning I put things inside `trulens.experimenta.otel_tracing.core.session.py` things which belong to `trulens.core.session.py` and would be integrated there once `otel_tracing` is adopted and no longer experimental. Inside the duplicated paths I also include some duplicated class names but with underscores like `_TruSession` for things that will be integrated into `TruSession` after the experimental phase.

- Adds experimental "otel-tracing" flag to enable the following functionality.

- Tracing of method calls as spans.

- Inclusion of spans in Records. This includes both the new trulens spans as well as spans created by other otel tools and spans explicitly created by the user.

- Exporting spans to otel exporter.

## Important things not included in this PR:

- Cost tracking when otel-tracing is enabled. This is part of the old otel PR but will need work to update.

- Inclusion of remote spans.

- TruLens spans tracking by non-trulens tracers.

- Span typing/organization. This is part of another old PR that needs work to update.

## Known issues:

- Existing problems with async prevent the OTEL tracing from operating on some apps using async calls.

## Other details good to know for developers

Please include any other details of this change useful for _TruLens_ developers.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] New Tests
- [ ] This change includes re-generated golden test results
- [ ] This change requires a documentation update

----

> [!IMPORTANT]
> This PR adds experimental OpenTelemetry tracing support, introducing new classes and modifying existing ones to trace method calls, manage spans, and export them to an OTEL exporter.
> 
>   - **Behavior**:
>     - Adds `otel-tracing` flag for experimental OTEL tracing functionality.
>     - Traces method calls as spans and includes spans in records.
>     - Exports spans to OTEL exporter.
>   - **Core Changes**:
>     - Introduces `Span`, `SpanContext`, `Tracer`, and `TracerProvider` classes in `trace.py` for OTEL compatibility.
>     - Implements `OTELExportable` for span conversion to OTEL format.
>     - Adds `PhantomSpanRecordingContext` for context management.
>   - **Session Management**:
>     - Updates `_TruSession` in `session.py` to handle OTEL exporter setup.
>   - **Utilities**:
>     - Adds `wrap_awaitable` and `wrap_iterable` in `wrap.py` for wrapping asynchronous and iterable objects.
>   - **Testing**:
>     - Updates golden test files to reflect new tracing capabilities.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=truera%2Ftrulens&utm_source=github&utm_medium=referral)<sup> for e550caa535aa515688f01ddc4a6e0abd07111667. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: examples/experimental/export_dummy_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,268 @@
+{
```
Comment (sfc-gh-jreini) [inline examples/experimental/export_dummy_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.    session.experimental_enable_feature("otel_tracing")
</pre><p>Can you update the order here slightly to make it more clear?</p><p> </p>

<pre class="ql-syntax" spellcheck="false"># start a trulens session
session = TruSession()

# (optional) if using another exporter to view traces, start the session with an otel exporter
session = TruSession(_experimental_otel_exporter=exporter)

# enable otel experimental feature: ## review note: even though this is done by default if the exporter is set above, it's clearer to have this laid out stepwise
session.experimental_enable_feature("otel_tracing")


</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1485/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (sfc-gh-jreini) [inline examples/experimental/export_dummy_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    # - ["trulens.bindings"] - Attributes specific to TruLens spans. Call spans include
</pre><p>what attributes are available here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1485/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (sfc-gh-jreini) [inline examples/experimental/export_dummy_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.        # Another custom span.
</pre><p>add (optional) to comment</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1485/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (sfc-gh-pmardziel) [reply examples/experimental/export_dummy_example.ipynb:1]: Changed to use a property to set the exporter so I think this is fixed:

```python
# Create a TruLens session.
session = TruSession()

# Set the exporter.
session.experimental_otel_exporter = exporter

# (Optional) Enable otel_tracing. Note that this is not required if you set the
# exporter above. If you would like to trace using spans without an exporter,
# this step is required.
session.experimental_enable_feature("otel_tracing")
```

File: src/core/trulens/experimental/otel_tracing/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# OTEL Tracing experimental feature code
+
+The folder structure from here is meant to correspond to trulens folder
+structure and if the experimental feature were to be adopted, the contents of
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/experimental/otel_tracing/README.md:4]: fixed

File: pyproject.toml [hunk 3]
```diff
@@ -28,7 +28,11 @@ classifiers = [
 
 [tool.poetry.dependencies]
 python = "^3.9,!=3.9.7"
-trulens-core = { version = "^1.0.0" }
+trulens-core = { version = "^1.0.0", extras = [
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply pyproject.toml:31]: I'm unsure what other group means here? Do you mean in some other package? or do you mean otel and experimental otel_tracing should be two different extras groups?

File: src/core/trulens/core/feedback/feedback.py [hunk 4]
```diff
@@ -666,6 +666,11 @@ def check_selectors(
 
         # with c.capture() as cap:
         for k, q in self.selectors.items():
+            if q.path[0:2] == Select.RecordSpans.path[0:2]:
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/feedback/feedback.py:669]: Made a utility method with more informative name. This is checking whether the given Lens q has the prefix that would select the new spans inside a record. Those cannot be predicted ahead of time which this method is checking so we must ignore those.

File: src/core/trulens/core/schema/record.py [hunk 5]
```diff
@@ -154,6 +164,32 @@ def __str__(self):
...
+        """Add a call to the record.
+
+        This is a method to make sure to add the call in a specific position sorted by endtime.
+        """
+
+        if len(self.calls) == 0:
+            self.calls.append(call)
+            return
+
+        for i, icall in enumerate(self.calls):
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/schema/record.py:177]: No so I had to fix.

File: src/core/trulens/core/schema/record.py [hunk 6]
```diff
@@ -154,6 +164,32 @@ def __str__(self):
     as the app that generated this record via `layout_calls_as_app`.
     """
 
+    def add_call(self, call: RecordAppCall):
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/schema/record.py:167]: Deleted this method and added more notes around where the same invariant is being maintained.

File: examples/experimental/otel/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# python deps, OTEL:
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply examples/experimental/otel/requirements.txt:1]: People wishing to run the example in this folder. This is demonstrating how to configure OTEL in different ways outside of python. Still testing though.

File: src/core/trulens/core/utils/json.py [hunk 8]
```diff
@@ -323,12 +323,12 @@ def recur(o):
 
         content = forward_value
 
+    elif isinstance(obj, Lens):  # special handling of paths
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/utils/json.py:326]: The highlighted line was just a typo/bug, it was a line in the wrong place. The other change regarding computed fields. I wanted to dump pydantic fields that are defined using the @computed_field which is something that can decorate properties to notify pydantic for special handling of properties.

File: src/core/trulens/core/utils/python.py [hunk 9]
```diff
@@ -1296,4 +1335,65 @@ def delete_singleton(self):
...
+
+    # Hold singleton instances here.
+    _instances: Dict[Hashable, SingletonInfo[Singleton]] = {}
+
+    def warning(self):
+        """Issue warning that this singleton already exists."""
+
+        k = self.__class__.__name__
+        if k in Singleton._instances:
+            Singleton._instances[k].warning()
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/utils/python.py:1354]: Instances contains `SingletonInfo` instances which have a different `warning` method. .

File: tests/utils.py [hunk 10]
```diff
@@ -379,6 +379,15 @@ def get_class_members(
 
         is_def = False
         is_public = False
+        is_experimental = False
+
+        if name.lower().startswith("experimental"):
+            is_experimental = True
+
+        if is_experimental:
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply tests/utils.py:387]: Did one of these things.

File: src/core/trulens/core/utils/pyschema.py [hunk 11]
```diff
@@ -426,8 +426,30 @@ class Bindings(SerialModel):
...
+                kwargs={
+                    k: v
+                    for k, v in b.arguments.items()
+                    if (not skip_self or k != "self")
+                },
+            )
+
+        if skip_self:
+            if "self" in b.arguments:
+                args = b.args[1:]
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/utils/pyschema.py:446]: I think it has to be the first arg but it doesn't need to be named "self" which is a problem here.

[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/utils/pyschema.py:446]: Changed it to be slightly more robust in terms of the possible names of the first arg.

File: src/core/trulens/experimental/otel_tracing/core/otel.py [hunk 12]
```diff
@@ -0,0 +1,738 @@
...
+
+def flatten_value(
+    v: TLensedAttributeValue, path: Optional[Lens] = None
+) -> Iterable[Tuple[Lens, types_api.AttributeValue]]:
+    """Flatten a lensed value into OpenTelemetry attribute values."""
+
+    if path is None:
+        path = Lens()
+
+    # if v is None:
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/experimental/otel_tracing/core/otel.py:97]: These particular lines are for when path is not given as input. The comment is a placeholder to remind us we need to do something about None values.

[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/experimental/otel_tracing/core/otel.py:97]: Also added more docs around related types like `TLensedAttributes` above this method.

File: src/core/trulens/core/utils/imports.py [hunk 13]
```diff
@@ -570,7 +570,11 @@ def __import__(self, name, globals=None, locals=None, fromlist=(), level=0):
         while frame.f_code == self.__import__.__code__:
             frame = frame.f_back
 
-        module_name = frame.f_globals["__name__"]
+        if "__name__" in frame.f_globals:
+            module_name = frame.f_globals["__name__"]
+        else:
+            # TODO: Figure out why we are even getting here.
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/utils/imports.py:576]: Don't want to raise here, I just want the next condition in that method to be False.

File: examples/experimental/otel/otel-collector-config.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
+# https://opentelemetry.io/docs/languages/python/getting-started/
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply examples/experimental/otel/otel-collector-config.yaml:1]: Added some comments. No check needed right now.

File: examples/experimental/otel/Makefile [hunk 15]
```diff
@@ -0,0 +1,44 @@
+PYTHON:=python3
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply examples/experimental/otel/Makefile:1]: added

File: examples/experimental/otel/Makefile [hunk 16]
```diff
@@ -0,0 +1,45 @@
+POETRY:=poetry run
+PYTHON:=python
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply examples/experimental/otel/Makefile:2]: Removing this entire folder from this PR. Will revisit later in another PR.

File: examples/experimental/otel/dummy.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
+# ruff: noqa: F401
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply examples/experimental/otel/dummy.py:1]: Removing this entire folder from this PR. Will revisit later in another PR.

File: src/core/trulens/core/instruments.py [hunk 18]
```diff
@@ -353,8 +353,17 @@ def finish_record(
                 for call in existing_record.calls:
                     current_calls[call.call_id] = call
 
+            # Maintain an order in a record' calls to make sure the root call
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmardziel) [reply src/core/trulens/core/instruments.py:356]: Fixed.

**Review by sfc-gh-dkurokawa — APPROVED** [review]

File: src/core/trulens/experimental/otel_tracing/core/otel.py [hunk 19]
```diff
@@ -0,0 +1,778 @@
...
+    See `TLensedAttributes` for more details.
+    """
+
+    if lens is None:
+        lens = Lens()
+
+    # OpenTelemetry does not allow None as an attribute value. Unsure what
+    # is best to do here.
+    # if v is None:
+    #    yield (path, "None")
```
Comment (sfc-gh-pdharmana) [inline src/core/trulens/experimental/otel_tracing/core/otel.py:123]: Please file a TODO jiras for these under otel epic

Comment (sfc-gh-pmardziel) [reply src/core/trulens/experimental/otel_tracing/core/otel.py]: Did this and some others.

**Pull request merged**